fetch-in-react-practice/client/src/components/Todos.tsx

```diff
@@ -0,0 +1,153 @@
+import { useEffect, useState } from 'react';
+import PageTitle from './PageTitle';
+import TodoList from './TodoList';
+import TodoForm from './TodoForm';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export default function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function fetchTodos() {
+      try {
+        const response = await fetch('/api/todos');
+        if (!response.ok) {
+          throw new Error(`Response invalid: ${response.status}`);
+        }
+        const fetchResponse = await response.json();
+        setTodos(fetchResponse);
+      } catch (error: any) {
+        console.error(error.message);
+        setError(error);
+      } finally {
+        setIsLoading(false);
+      }
+    }
+    fetchTodos();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    try {
+      const response = await fetch('/api/todos', {
+        method: 'POST',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify(newTodo),
+      });
+      if (!response.ok) {
+        throw new Error(`Response invalid: ${response.status}`);
+      }
+      const fetchResponse = await response.json();
+      setTodos([...todos, fetchResponse]);
+    } catch (error: any) {
+      console.error(error.message);
+      setError(error);
+    }
+  }
+
+  /* Implement toggleCompleted to toggle the completed state of a todo. Hints are at the bottom of the file. */
+  async function toggleCompleted(todoId: number) {
+    const todoIndex = todos.findIndex((x) => x.todoId === todoId);
+    if (!todoIndex) throw new Error('Todo not found with that ID');
```

Comment:
This isn't actually correct. Since todoIndex is a number, `!todoIndex` will be truthy if todoIndex is 0 and falsey otherwise. This means you cannot find the first value. Try toggling the first item in your list -- you won't be able to!

`findIndex` returns -1 if it can't find the item, so this check should be `if (todoIndex < 0)`.